### PR TITLE
refactor: Customise enhanced meshing workflow

### PR DESCRIPTION
**Pull request by prmukherj**
Clean-up of test.
Fixing **insert** to work with python names.
Object orientation of **insert**.
Loading a saved workflow.
Creating new workflows.

### CODE REVIEW

File: src/ansys/fluent/core/workflow.py [hunk 1]
```diff
@@ -504,19 +526,52 @@ def insert_next_task(self, command_name: str):
         Raises
         ------
         ValueError
-            If the command name does not match a task name.
+            If the python name does not match the next possible task names.
```
Comment (PipKat) [inline src/ansys/fluent/core/workflow.py:529]:
```suggestion
            If the Python name does not match the next possible task names.
```

File: src/ansys/fluent/core/workflow.py [hunk 2]
```diff
@@ -504,19 +526,52 @@ def insert_next_task(self, command_name: str):
...
         """
-        if command_name not in self.get_next_possible_tasks():
+        # The next line populates the python name map for next possible task
```
Comment (PipKat) [inline src/ansys/fluent/core/workflow.py:531]:
```suggestion
        # The next line populates the Python name map for next possible task
```

File: src/ansys/fluent/core/workflow.py [hunk 3]
```diff
@@ -504,19 +526,52 @@ def insert_next_task(self, command_name: str):
...
         return self._task.InsertNextTask(
-            CommandName=snake_to_camel_case(
-                command_name, self._task.GetNextPossibleTasks()
-            )
+            CommandName=self._python_task_names_map[task_name]
         )
 
+    @property
+    def next_tasks(self):
+        """Tasks that can be inserted after this current task."""
```
Comment (PipKat) [inline src/ansys/fluent/core/workflow.py:544]:
```suggestion
        """Tasks that can be inserted after the current task."""
```

File: src/ansys/fluent/core/workflow.py [hunk 4]
```diff
@@ -504,19 +526,52 @@ def insert_next_task(self, command_name: str):
...
+            return self._insertable_tasks
+
+        class _Insert:
+            def __init__(self, base_task, name):
+                """Initialize _Insert."""
+                self._base_task = base_task
+                self._name = name
+
+            def insert(self):
+                """Inserts a task in the workflow."""
```
Comment (PipKat) [inline src/ansys/fluent/core/workflow.py:569]:
```suggestion
                """Insert a task in the workflow."""
```

**Review by hpohekar — APPROVED** [review]

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 5]
```diff
@@ -240,3 +240,67 @@ class WorkflowMode(Enum):
...
+    """Provides a specialization of the workflow wrapper for a loaded workflow."""
+
+    def __init__(
+        self,
+        workflow: PyMenuGeneric,
+        meshing: PyMenuGeneric,
+        file_path: str,
+        fluent_version: FluentVersion,
+    ) -> None:
+        """Initialize LoadWorkflow.
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:255]:
```suggestion
        """Initialize a ``LoadWorkflow`` instance.
```

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 6]
```diff
@@ -240,3 +240,67 @@ class WorkflowMode(Enum):
...
+        self._meshing = meshing
+        self._file_path = file_path
+
+    def load(self) -> None:
+        """Load a workflow."""
+        self._load_workflow(file_path=self._file_path)
+
+
+class CreateWorkflow(Workflow):
+    """Provides a specialization of the workflow wrapper for newly created workflow."""
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:280]:
```suggestion
    """Provides a specialization of the workflow wrapper for a newly created workflow."""
```

File: src/ansys/fluent/core/meshing/meshing_workflow.py [hunk 7]
```diff
@@ -240,3 +240,67 @@ class WorkflowMode(Enum):
...
+
+    def __init__(
+        self,
+        workflow: PyMenuGeneric,
+        meshing: PyMenuGeneric,
+        fluent_version: FluentVersion,
+    ) -> None:
+        """Initialize CreateWorkflow.
```
Comment (PipKat) [inline src/ansys/fluent/core/meshing/meshing_workflow.py:288]:
```suggestion
        """Initialize a ``CreateWorkflow`` instance.
```

File: src/ansys/fluent/core/session_base_meshing.py [hunk 8]
```diff
@@ -199,6 +205,28 @@ def topology_based_meshing_workflow(self):
...
+                self._workflow_se,
+                self.meshing,
+                file_path,
+                self.get_fluent_version(),
+            )
+        return self._loaded_workflow
+
+    @property
+    def create_workflow(self):
+        """Datamodel root of workflow exposed in object-oriented manner."""
```
Comment (PipKat) [inline src/ansys/fluent/core/session_base_meshing.py:221]:
```suggestion
        """Datamodel root of the workflow exposed in an object-oriented manner."""
```

File: src/ansys/fluent/core/session_base_meshing.py [hunk 9]
```diff
@@ -199,6 +205,28 @@ def topology_based_meshing_workflow(self):
...
+            self._created_workflow = CreateWorkflow(
+                self._workflow_se,
+                self.meshing,
+                self.get_fluent_version(),
+            )
+        return self._created_workflow
+
     @property
     def PartManagement(self):
         """Datamdoel root of PartManagement."""
```
Comment (PipKat) [inline src/ansys/fluent/core/session_base_meshing.py:232]:
```suggestion
        """Datamodel root of PartManagement."""
```
PartManagement should be in double backticks to format it as a code entity and a noun should follow that to indicate what type of entity it is.

File: src/ansys/fluent/core/session_pure_meshing.py [hunk 10]
```diff
@@ -111,6 +111,16 @@ def two_dimensional_meshing(self):
         self._base_meshing.two_dimensional_meshing_workflow.reinitialize()
         return self._base_meshing.two_dimensional_meshing_workflow
 
+    def load_workflow(self, file_path: str):
+        """Load a saved workflow."""
+        self._base_meshing.load_workflow(file_path=file_path).load()
+        return self._base_meshing.load_workflow(file_path=file_path)
+
+    def create_workflow(self):
+        """Create a new meshing workflow."""
```
Comment (PipKat) [inline src/ansys/fluent/core/session_pure_meshing.py:120]:
```suggestion
        """Create a meshing workflow."""
```
The "new" is unnecessary as you can't create an "old" workflow.

File: src/ansys/fluent/core/session_pure_meshing.py [hunk 11]
```diff
@@ -111,6 +111,16 @@ def two_dimensional_meshing(self):
...
+        self._base_meshing.create_workflow.create()
+        return self._base_meshing.create_workflow
+
     def topology_based(self):
         """Get a new topology-based meshing workflow.
```
Comment (PipKat) [inline src/ansys/fluent/core/session_pure_meshing.py:125]: Should the be "Get a new ...." as it is now or "Create a topology-based ...."?

File: src/ansys/fluent/core/workflow.py [hunk 12]
```diff
@@ -504,19 +527,52 @@ def insert_next_task(self, command_name: str):
...
+            If the Python name does not match the next possible task names.
         """
-        if command_name not in self.get_next_possible_tasks():
+        # The next line populates the Python name map for next possible task
+        self._get_next_python_task_names()
+        if task_name not in self.get_next_possible_tasks():
             raise ValueError(
-                f"'{command_name}' cannot be inserted next to '{self.python_name()}'. \n"
+                f"'{task_name}' cannot be inserted next to '{self.python_name()}'. \n"
                 "Please use 'get_next_possible_tasks()' to view list of allowed tasks."
```
Comment (PipKat) [inline src/ansys/fluent/core/workflow.py:537]:
```suggestion
                "Use 'get_next_possible_tasks()' to view list of allowed tasks."
```
Omit "please" in technical doc.

File: src/ansys/fluent/core/workflow.py [hunk 13]
```diff
@@ -504,19 +527,52 @@ def insert_next_task(self, command_name: str):
...
         )
 
+    @property
+    def next_tasks(self):
+        """Tasks that can be inserted after the current task."""
+        return self._NextTask(self)
+
+    class _NextTask:
+        def __init__(self, base_task):
+            """Initialize _NextTask."""
```
Comment (PipKat) [inline src/ansys/fluent/core/workflow.py:550]:
```suggestion
            """Initialize an ``_NextTask`` instance."""
```

File: src/ansys/fluent/core/workflow.py [hunk 14]
```diff
@@ -1398,11 +1457,82 @@ def _attr_from_wrapped_workflow(self, attr):
...
+        self._workflow.InitializeWorkflow(WorkflowType=name)
+        self._activate_dynamic_interface(dynamic_interface=dynamic_interface)
+
+    def _load_workflow(self, file_path: str, dynamic_interface: bool = True):
+        self._workflow.LoadWorkflow(FilePath=file_path)
+        self._activate_dynamic_interface(dynamic_interface=dynamic_interface)
+
+    def get_initial_task_list_while_creating_new_workflow(self):
+        """Get list of independent tasks that can be inserted at initial level while
+        creating a new workflow."""
```
Comment (PipKat) [inline src/ansys/fluent/core/workflow.py:1474]:
```suggestion
        """Get a list of independent tasks that can be inserted at the initial level when
        creating a workflow."""
```

File: src/ansys/fluent/core/workflow.py [hunk 15]
```diff
@@ -1398,11 +1457,82 @@ def _attr_from_wrapped_workflow(self, attr):
...
+        self._activate_dynamic_interface(dynamic_interface=dynamic_interface)
+
+    @property
+    def first_tasks(self):
+        """Tasks that can be inserted on a blank workflow."""
+        return self._FirstTask(self)
+
+    class _FirstTask:
+        def __init__(self, workflow):
+            """Initialize _FirstTask."""
```
Comment (PipKat) [inline src/ansys/fluent/core/workflow.py:1489]:
```suggestion
            """Initialize an ``_FirstTask`` instance."""
```

File: src/ansys/fluent/core/session_pure_meshing.py (hunk 11 above)
Comment (prmukherj) [reply src/ansys/fluent/core/session_pure_meshing.py:125]: yes, it should be get as it returns a created workflow.

**Pull request merged**